### ISSUE TITLE
GHC backend: Constructor argument is erased in spite of COMPILE data pragma

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#3732]
Working on BNFC's Agda backend I discovered the following problem.
```agda
data I : Set where
  bar : I

{-# COMPILE GHC I = data I (Bar) #-}
```
The generated Haskell module 
```haskell
name2 = "Bug.I"
d2 = ()
type T2 = I
pattern C4 = Bar
check4 :: T2
check4 = Bar
cover2 :: I -> ()
cover2 x
  = case x of
      Bar -> ()a
```
has the following issues:
```
$ agda -c --no-main Bug.agda
Calling: ghc -O -Werror -iBug.hs --make -fwarn-incomplete-patterns -fno-warn-overlapping-patterns
[2 of 2] Compiling MAlonzo.Code.Bug ( ... )
Compilation error:
    Not in scope: type constructor or class ‘I’
   |
14 | type T2 = I
   |           ^

    Not in scope: data constructor ‘Bar’
   |
15 | pattern C4 = Bar
   |              ^^^

    Not in scope: type constructor or class ‘I’
   |
18 | cover2 :: I -> ()
   |           ^

    Not in scope: data constructor ‘Bar’
   |
21 |       Bar -> ()
   |       ^^^
```

**Comment by nad** [comment]
This looks correct to me.

**Comment by UlfNorell** [comment]
You haven't defined the datatype on the Haskell side as far as I can see. Were you expecting it to define it for you?

**Comment by andreasabel** [comment]
Ah, I shrank too much.

**Comment by andreasabel** [comment]
```agda
{-# FOREIGN GHC data I = Bar #-}
{-# FOREIGN GHC data S = Foo I #-}

data I : Set where
  bar : I

{-# COMPILE GHC I = data I (Bar) #-}

data S : Set where
  foo : (i : I) → S

{-# COMPILE GHC S = data S (Foo) #-}
```
This reports:
```Compilation error:

Bug.hs:27:15: error:
    • The constructor ‘Foo’ should have 1 argument, but has been given none
    • In the pattern: Foo
      In the declaration for pattern synonym ‘C10’
   |
27 | pattern C10 = Foo
   |               ^^^

Bug.hs:29:11: error:
    • Couldn't match type ‘I -> S’ with ‘S’
      Expected type: T6
        Actual type: I -> S
    • Probable cause: ‘Foo’ is applied to too few arguments
      In the expression: Foo
      In an equation for ‘check10’: check10 = Foo
   |
29 | check10 = Foo
   |           ^^^

Bug.hs:33:7: error:
    • The constructor ‘Foo’ should have 1 argument, but has been given none
    • In the pattern: Foo
      In a case alternative: Foo -> ()
      In the expression: case x of { Foo -> () }
   |
33 |       Foo -> ()
   |       ^^^
```
The compiled file is:
```haskell

data S = Foo I
data I = Bar
name2 = "Bug.I"
d2 = ()
type T2 = I
pattern C4 = Bar
check4 :: T2
check4 = Bar
cover2 :: I -> ()
cover2 x
  = case x of
      Bar -> ()
name6 = "Bug.S"
d6 = ()
type T6 = S
pattern C10 = Foo
check10 :: T6
check10 = Foo
cover6 :: S -> ()
cover6 x
  = case x of
      Foo -> ()
```
It seems that the erasure machinery runs on the constructor `foo` even though its translation to Haskell is fixed by the `COMPILE` pragma.

**Comment by UlfNorell** [comment]
Ah yes, that's because `I` gets erased. It might be possible to have the ghc backend tell the treeless compiler not to erase types with a `data` pragma.

**Comment by nad** [comment]
See #2921 for some previous discussion.

**Comment by andreasabel** [comment]
Even if I as a human could figure out that the argument to `foo` gets erased, a tool such that BNFC should not have to reimplement Agda's erasure machinery only to be able to generate Agda bindings for generated data types.

This is bringing me once more back to my mantra: __declare, not infer__.
The base line always needs to be the ability to declare what you want.  Inference is only if you are too lazy to write it out and the machine may fill it in for you.  Erasure, forcing, termination, positivity and what not should not be an exception to this principle.

> It might be possible to have the ghc backend tell the treeless compiler not to erase types with a data pragma.

We should also consider allowing the user to write the translation from Agda constructor to Haskell constructor write out in more detail, with the possibility of dropping (and reordering) arguments.

Another way to fix the problem at hand is to have treeless not erase arguments with @omega attribute.  

What is sure is that the current behavior is too brittle, the default should be that types with a pragma (which only lists the constructor names) should not be subject to erasure. The erasure should be visible to the reader of the code as well, which could be achieved with more detailed pragmas.

**Comment by xekoukou** [comment]
In agda-ocaml, I make the assumption that what you see is what you get. Of course, with erasure, this is not true. It would be true if we could disable erasure for a specific data type.

**Comment by andreasabel** [comment]
I pushed the fix I discussed with @UlfNorell: Every type that has a GHC COMPILE binding is treated as non-erasable.
The unwanted side effect is maybe that now the Builtin.Unit type is non-erasable.
We have to see what the effect on performance will be.

**Comment by andreasabel** [comment]
@xekoukou: The GHC backend uses the `Treeless` compiler for erasure.  I added a check in Treeless whether erasure for a data/record type is disallowed (new fields `dataErasureAllowed`/`recErasureAllowed` in `Signature`).  The permission is removed for types that have a `GHC COMPILE` binding.

Does your compiler also go through Treeless?

**Comment by xekoukou** [comment]
@andreasabel I think it works for me too. ( I use the ```toTreeless``` function.) I just need to add a data compile pragma so that it stops erasure for the specific data type.

**Comment by andreasabel** [comment]
My patch 53de2254459948846b2f895b040ae427f39ac5f6 broke the travis test: https://travis-ci.org/agda/agda/builds/527794823.
@erydo commented https://github.com/agda/agda/pull/3589#issuecomment-489335511 that the difference between golden vs. actual output is:
```diff
agda-tests: Call to `git diff` failed: (ExitFailure 1,"diff --git a/tmp/CompileAsPattern4333-0.golden b/tmp/CompileAsPattern4333-1.actual
index 105a09d..8736cec 100644
--- a/tmp/CompileAsPattern4333-0.golden
+++ b/tmp/CompileAsPattern4333-1.actual
@@ -16,8 +16,7 @@ out >               (Agda.Builtin.List.List._\8759_ f (CompileAsPattern._.merge _
 out > CompileAsPattern.mapM! =
 out >   \955 _ a b \8594
 out >     case b of
-out >       Agda.Builtin.List.List.[] \8594
-out >         Common.IO.return () _ Agda.Builtin.Unit.\8868.tt
+out >       Agda.Builtin.List.List.[] \8594 Common.IO.return () _ _
 out >       Agda.Builtin.List.List._\8759_ c d \8594
 out >         Common.IO._>>=_
 out >           () () _ _ (a c) (\955 _ \8594 CompileAsPattern.mapM! _ a d)
","")
```
On my machine, the `Agda.Builtin.Unit.\8868.tt` is actually produced instead of `_`, thus I put it into the patch.
```diff
diff --git a/test/Compiler/simple/CompileAsPattern.out b/test/Compiler/simple/CompileAsPattern.out
index 8736cecbc..105a09d4a 100644
--- a/test/Compiler/simple/CompileAsPattern.out
+++ b/test/Compiler/simple/CompileAsPattern.out
@@ -16,7 +16,8 @@ out >               (Agda.Builtin.List.List._∷_ f (CompileAsPattern._.merge _
 out > CompileAsPattern.mapM! =
 out >   λ _ a b →
 out >     case b of
-out >       Agda.Builtin.List.List.[] → Common.IO.return () _ _
+out >       Agda.Builtin.List.List.[] →
+out >         Common.IO.return () _ Agda.Builtin.Unit.⊤.tt
 out >       Agda.Builtin.List.List._∷_ c d →
 out >         Common.IO._>>=_
 out >           () () _ _ (a c) (λ _ → CompileAsPattern.mapM! _ a d)
```
I am mystified why travis behaves differently here.

@erydo: the diff you report is from travis, or?  Do you get the same problem on your machine or is it only on travis?

**Comment by asr** [comment]
After seeing the `.travis.yml` file, I couldn't reproduce the Travis error running

```
$ export PARALLEL_TESTS=2
$ make AGDA_TESTS_OPTIONS="-j${PARALLEL_TESTS} --hide-successes" compiler-test
```

**Comment by rwe** [comment]
The diff I posted was from Travis, I haven't tried to repro locally yet

**Comment by rwe** [comment]
It'll take me a while to get everything building locally, but I'm working on it. Since @UlfNorell mentioned it did not fail on his machine either: Could this be a caching issue? Like the `.agdai` being generated and then reused, but behaving differently?

**Comment by rwe** [comment]
Here's a more specific idea (cc @andreasabel):
  * Multiple tests import `test/Common/Prelude/*`
  * Those generate `.agdai` files.
  * These tests are run in parallel.
  * In 53de2254459948846b2f895b040ae427f39ac5f6, types default to "erasable" unless a binding indicates otherwise.
  * The specific set of imports determines which bindings are encountered.
  * The erasability of a type is [serialized/deserialized](https://github.com/agda/agda/commit/53de2254459948846b2f895b040ae427f39ac5f6#diff-a17dbbaf803bf8e5a0ff088af21fc2baR324)
  * Therefore, which bindings are considered erasable in the cached `test/Common/Prelude/**/*.agdai` files might differ depending on which test caused the `.agdai` file to be created.

Seem plausible?

**Comment by rwe** [comment]
Basically, then, you should be able to reproduce this by a test scenario something like:

*`Shared.agda`*
```
-- Some function using a type like ⊤
```

*`IsNonErasable.agda`*
```agda
open import Shared
-- Bind ⊤ to a Haskell symbol with `FOREIGN` and `COMPILE`.
```

*`IsErasable.agda`*
```agda
open import Shared
-- Not binding ⊤
```

If that hypothesis is right, then deleting all `.agdai` files and running `IsErasable` would show `_`. But if you run it after `IsNonErasable`, it would erroneously show `⊤`. (Or something like that). The behaviour would be dependent on which invocation created `Shared.agdai`.

Erasability doesn't seem like it should be cached this way (ie the per-module `.agdai`). Whether something can be erased depends on proving that nothing else uses it or will use it. That determination can only be done for a given full compilation set, and doesn't transfer to other possible compilation sets.

**Comment by rwe** [comment]
#3752 (deleting the `.agdai` files beforehand running the Compiler tests) failed with the same error: https://travis-ci.org/agda/agda/jobs/529020401#L4924

Though, if it's a problem _between_ the Compiler tests, it could still be the cause. So my next suggestion would be to disable parallelism in those tests and see if they behave differently.

**Comment by rwe** [comment]
Well, that doesn't seem to be working either. But I have two more good clues CC @andreasabel.

If I executing the test compilation manually, locally:
```sh
# The following arguments are the same ones the tests run.
# (except that --compile-dir is a temp directory for the tests).
# The printf subshell at the bottom is to format it the same as the .out file for diffing.
stack exec "${AGDA_BIN}" -- \
    --compile \
    --ignore-interfaces \
    --no-libraries \
    --compile-dir . \
    -v0 \
    -vwarning:1 \
    -itest/Compiler/simple \
    -itest/ \
    --ghc-flag=-itest/ \
    test/Compiler/simple/CompileAsPattern.agda \
  | ( \
      printf 'EXECUTED_PROGRAM\n\nret > ExitSuccess\n'; \
      sed 's/^/out > /g'; \
      printf 'out > 012235810101520\n'; \
  ) > test/Compiler/simple/CompileAsPattern.real
```

And diffing:
```sh
git diff --no-index test/Compiler/simple/CompileAsPattern.{out,real}
```

Exhibits the problem locally. This does not happen if I just run the tests via `stack build --test`.
```diff
diff --git a/test/Compiler/simple/CompileAsPattern.out b/test/Compiler/simple/CompileAsPattern.real
index 105a09d4a..8736cecbc 100644
--- a/test/Compiler/simple/CompileAsPattern.out
+++ b/test/Compiler/simple/CompileAsPattern.real
@@ -16,8 +16,7 @@ out >               (Agda.Builtin.List.List._∷_ f (CompileAsPattern._.merge _
 out > CompileAsPattern.mapM! =
 out >   λ _ a b →
 out >     case b of
-out >       Agda.Builtin.List.List.[] →
-out >         Common.IO.return () _ Agda.Builtin.Unit.⊤.tt
+out >       Agda.Builtin.List.List.[] → Common.IO.return () _ _
 out >       Agda.Builtin.List.List._∷_ c d →
 out >         Common.IO._>>=_
 out >           () () _ _ (a c) (λ _ → CompileAsPattern.mapM! _ a d)
```

**Also interesting and probably relevant**: If I run `"${AGDA_BIN}"` directly — i.e. the same command as above, but removing the `stack exec --` — I get partial compilation (_without having erased the symbol as above_) and then an error.

(I think the error is just a module search path thing, not actually relevant).
```
out > CompileAsPattern._.merge =
out >   λ _ _ a b c →
out >     case b of
out >       Agda.Builtin.List.List.[] → c
out >       Agda.Builtin.List.List._∷_ d e →
out >         case c of
out >           Agda.Builtin.List.List.[] → b
out >           Agda.Builtin.List.List._∷_ f g →
out >             Common.Bool.if_then_else_
out >               _ _ (a d f)
out >               (Agda.Builtin.List.List._∷_ d (CompileAsPattern._.merge _ _ a e c))
out >               (Agda.Builtin.List.List._∷_ f (CompileAsPattern._.merge _ _ a b g))
out > CompileAsPattern.mapM! =
out >   λ _ a b →
out >     case b of
out >       Agda.Builtin.List.List.[] →
out >         Common.IO.return () _ Agda.Builtin.Unit.⊤.tt
out >       Agda.Builtin.List.List._∷_ c d →
out >         Common.IO._>>=_
out >           () () _ _ (a c) (λ _ → CompileAsPattern.mapM! _ a d)
out > CompileAsPattern.xs =
out >   Agda.Builtin.List.List._∷_
out >     2
out >     (Agda.Builtin.List.List._∷_
out >        3
out >        (Agda.Builtin.List.List._∷_
out >           5
out >           (Agda.Builtin.List.List._∷_
out >              10 (Agda.Builtin.List.List._∷_ 20 Agda.Builtin.List.List.[]))))
out > CompileAsPattern.ys =
out >   Agda.Builtin.List.List._∷_
out >     0
out >     (Agda.Builtin.List.List._∷_
out >        1
out >        (Agda.Builtin.List.List._∷_
out >           2
out >           (Agda.Builtin.List.List._∷_
out >              8
out >              (Agda.Builtin.List.List._∷_
out >                 10 (Agda.Builtin.List.List._∷_ 15 Agda.Builtin.List.List.[])))))
out > CompileAsPattern.main =
out >   CompileAsPattern.mapM!
out >     _ Common.IO.printNat
out >     (CompileAsPattern._.merge
out >        _ _ _<_ CompileAsPattern.xs CompileAsPattern.ys)
out > Compilation error:
out >
out > MAlonzo/RTE.hs:8:1: error:
out >     Could not find module ‘Numeric.IEEE’
out >     Use -v to see a list of the files searched for.
out >   |
out > 8 | import Numeric.IEEE ( IEEE(identicalIEEE) )
out >   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
out >
out > 012235810101520
```

The diff of _that_ output against the golden `.out` is just the bottom error; everything else is "correct":
```diff
index 105a09d4a..49c273c09 100644
--- a/test/Compiler/simple/CompileAsPattern.out
+++ b/test/Compiler/simple/CompileAsPattern.real
@@ -46,4 +46,13 @@ out >   CompileAsPattern.mapM!
 out >     _ Common.IO.printNat
 out >     (CompileAsPattern._.merge
 out >        _ _ _<_ CompileAsPattern.xs CompileAsPattern.ys)
+out > Compilation error:
+out >
+out > MAlonzo/RTE.hs:8:1: error:
+out >     Could not find module ‘Numeric.IEEE’
+out >     Use -v to see a list of the files searched for.
+out >   |
+out > 8 | import Numeric.IEEE ( IEEE(identicalIEEE) )
+out >   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+out >
 out > 012235810101520
```

Which makes me wonder what environmental factors are influence the compiler invocations during the tests.

N.B. I use stack to build and set `AGDA_BIN` like:
```sh
export AGDA_BIN="$(stack path --local-install-root)/bin/agda"
```

**Comment by rwe** [comment]
After adding `--ghc-flag=-j2` to the second command above, it began to _also_ eliminate the `⊤` symbol. (In other words, failing the test). Removing the flag restored the `⊤`.

Though, I just installed `ieee` globally to fix that error, and now running manually eliminates the `⊤` in all cases, just like `stack exec` did; they're behaving the same now. So my current state is: `stack exec ${AGDA_BIN} -- …` and `${AGDA_BIN} …` run identically (producing a "failing" output). `stack build --test` passes the tests.

**Comment by rwe** [comment]
Ok, still not sure why — but if I do:
```sh
rm -rf CompileAsPattern MAlonzo/ && "${AGDA_BIN}" --compile …etc… 
```
(The command from my comment), the output equals `CompileAsPattern.out`.

If do *not* remove that output first, and just run `"${AGDA_BIN}" --compile …`, it generates something different.

Diffing the output itself, i.e. initially versus the output files created *if they already existed*, there's one difference, which corresponds closely to the test output:
```diff
diff --git a/MAlonzo/Code/CompileAsPattern.hs b/MAlonzo/Code/CompileAsPattern.hs
index 857aabf8d..70e636691 100644
--- a/MAlonzo/Code/CompileAsPattern.hs
+++ b/MAlonzo/Code/CompileAsPattern.hs
@@ -60,10 +60,7 @@ du32 ::
     AgdaAny MAlonzo.Code.Agda.Builtin.Unit.T6
 du32 v0 v1
   = case coe v1 of
-      []
-        -> coe
-             MAlonzo.Code.Common.IO.d8 () erased
-             MAlonzo.Code.Agda.Builtin.Unit.C8
+      [] -> coe MAlonzo.Code.Common.IO.d8 () erased erased
       (:) v2 v3
         -> coe
              MAlonzo.Code.Common.IO.d18 () () erased erased (coe v0 v2)
```

`rm -v MAlonzo/**/*.{hi,o}` does not count as clean. `rm -v MAlonzo/**/*.hs` does. So something is probably looking for `.hs` files existing in the output and then skipping some processing that it shouldn't.

That's as far as I can take this one for now.

**Comment by rwe** [comment]
Ok, I think this it's caused by something around here: https://github.com/agda/agda/blob/b8473159f43720932e1f96e90cd01474893b89c5/src/full/Agda/Compiler/MAlonzo/Compiler.hs#L155-L183

Running cleanly (which produces matching output) and with `-vcompile.ghc:2` gives me:
```
out > Compiling Agda.Primitive in ./agda/.stack-work/install/x86_64-osx/lts-13.2/8.6.3/share/x86_64-osx-ghc-8.6.3/Agda-2.6.1/lib/prim/Agda/Primitive.agdai to ./MAlonzo/Code/Agda/Primitive.hs
out > Compiling Common.Level in ./agda/test/Common/Level.agdai to ./MAlonzo/Code/Common/Level.hs
out > Compiling Agda.Builtin.List in ./agda/.stack-work/install/x86_64-osx/lts-13.2/8.6.3/share/x86_64-osx-ghc-8.6.3/Agda-2.6.1/lib/prim/Agda/Builtin/List.agdai to ./MAlonzo/Code/Agda/Builtin/List.hs
out > Compiling Agda.Builtin.Bool in ./agda/.stack-work/install/x86_64-osx/lts-13.2/8.6.3/share/x86_64-osx-ghc-8.6.3/Agda-2.6.1/lib/prim/Agda/Builtin/Bool.agdai to ./MAlonzo/Code/Agda/Builtin/Bool.hs
out > Compiling Agda.Builtin.Nat in ./agda/.stack-work/install/x86_64-osx/lts-13.2/8.6.3/share/x86_64-osx-ghc-8.6.3/Agda-2.6.1/lib/prim/Agda/Builtin/Nat.agdai to ./MAlonzo/Code/Agda/Builtin/Nat.hs
out > Compiling Common.Nat in ./agda/test/Common/Nat.agdai to ./MAlonzo/Code/Common/Nat.hs
out > Compiling Common.List in ./agda/test/Common/List.agdai to ./MAlonzo/Code/Common/List.hs
out > Compiling Agda.Builtin.Unit in ./agda/.stack-work/install/x86_64-osx/lts-13.2/8.6.3/share/x86_64-osx-ghc-8.6.3/Agda-2.6.1/lib/prim/Agda/Builtin/Unit.agdai to ./MAlonzo/Code/Agda/Builtin/Unit.hs
out > Compiling Common.Unit in ./agda/test/Common/Unit.agdai to ./MAlonzo/Code/Common/Unit.hs
out > Compiling Agda.Builtin.Char in ./agda/.stack-work/install/x86_64-osx/lts-13.2/8.6.3/share/x86_64-osx-ghc-8.6.3/Agda-2.6.1/lib/prim/Agda/Builtin/Char.agdai to ./MAlonzo/Code/Agda/Builtin/Char.hs
out > Compiling Agda.Builtin.String in ./agda/.stack-work/install/x86_64-osx/lts-13.2/8.6.3/share/x86_64-osx-ghc-8.6.3/Agda-2.6.1/lib/prim/Agda/Builtin/String.agdai to ./MAlonzo/Code/Agda/Builtin/String.hs
out > Compiling Agda.Builtin.Int in ./agda/.stack-work/install/x86_64-osx/lts-13.2/8.6.3/share/x86_64-osx-ghc-8.6.3/Agda-2.6.1/lib/prim/Agda/Builtin/Int.agdai to ./MAlonzo/Code/Agda/Builtin/Int.hs
out > Compiling Common.Integer in ./agda/test/Common/Integer.agdai to ./MAlonzo/Code/Common/Integer.hs
out > Compiling Common.Bool in ./agda/test/Common/Bool.agdai to ./MAlonzo/Code/Common/Bool.hs
out > Compiling Common.Char in ./agda/test/Common/Char.agdai to ./MAlonzo/Code/Common/Char.hs
out > Compiling Common.String in ./agda/test/Common/String.agdai to ./MAlonzo/Code/Common/String.hs
out > Compiling Agda.Builtin.Float in ./agda/.stack-work/install/x86_64-osx/lts-13.2/8.6.3/share/x86_64-osx-ghc-8.6.3/Agda-2.6.1/lib/prim/Agda/Builtin/Float.agdai to ./MAlonzo/Code/Agda/Builtin/Float.hs
out > Compiling Common.Float in ./agda/test/Common/Float.agdai to ./MAlonzo/Code/Common/Float.hs
out > Compiling Common.Maybe in ./agda/test/Common/Maybe.agdai to ./MAlonzo/Code/Common/Maybe.hs
out > Compiling Agda.Builtin.IO in ./agda/.stack-work/install/x86_64-osx/lts-13.2/8.6.3/share/x86_64-osx-ghc-8.6.3/Agda-2.6.1/lib/prim/Agda/Builtin/IO.agdai to ./MAlonzo/Code/Agda/Builtin/IO.hs
out > Compiling Common.IO in ./agda/test/Common/IO.agdai to ./MAlonzo/Code/Common/IO.hs
out > Compiling Common.Prelude in ./agda/test/Common/Prelude.agdai to ./MAlonzo/Code/Common/Prelude.hs
out > Compiling CompileAsPattern in ./agda/test/Compiler/simple/CompileAsPattern.agdai to ./MAlonzo/Code/CompileAsPattern.hs
```

Running again (which does not match the `.out`, and has erased the `Unit` type) gives me:
```
out > Agda.Primitive : no compilation is needed.
out > Compiling Common.Level in ./agda/test/Common/Level.agdai to ./MAlonzo/Code/Common/Level.hs
out > Agda.Builtin.List : no compilation is needed.
out > Agda.Builtin.Bool : no compilation is needed.
out > Agda.Builtin.Nat : no compilation is needed.
out > Compiling Common.Nat in ./agda/test/Common/Nat.agdai to ./MAlonzo/Code/Common/Nat.hs
out > Compiling Common.List in ./agda/test/Common/List.agdai to ./MAlonzo/Code/Common/List.hs
out > Agda.Builtin.Unit : no compilation is needed.
out > Compiling Common.Unit in ./agda/test/Common/Unit.agdai to ./MAlonzo/Code/Common/Unit.hs
out > Agda.Builtin.Char : no compilation is needed.
out > Agda.Builtin.String : no compilation is needed.
out > Agda.Builtin.Int : no compilation is needed.
out > Compiling Common.Integer in ./agda/test/Common/Integer.agdai to ./MAlonzo/Code/Common/Integer.hs
out > Compiling Common.Bool in ./agda/test/Common/Bool.agdai to ./MAlonzo/Code/Common/Bool.hs
out > Compiling Common.Char in ./agda/test/Common/Char.agdai to ./MAlonzo/Code/Common/Char.hs
out > Compiling Common.String in ./agda/test/Common/String.agdai to ./MAlonzo/Code/Common/String.hs
out > Agda.Builtin.Float : no compilation is needed.
out > Compiling Common.Float in ./agda/test/Common/Float.agdai to ./MAlonzo/Code/Common/Float.hs
out > Compiling Common.Maybe in ./agda/test/Common/Maybe.agdai to ./MAlonzo/Code/Common/Maybe.hs
out > Agda.Builtin.IO : no compilation is needed.
out > Compiling Common.IO in ./agda/test/Common/IO.agdai to ./MAlonzo/Code/Common/IO.hs
out > Compiling Common.Prelude in ./agda/test/Common/Prelude.agdai to ./MAlonzo/Code/Common/Prelude.hs
out > Compiling CompileAsPattern in ./agda/test/Compiler/simple/CompileAsPattern.agdai to ./MAlonzo/Code/CompileAsPattern.hs
```
CC @andreasabel

**Comment by andreasabel** [comment]
Thanks @erydo, for the excellent research!  I think I understand the problem now.
I think the following scenario triggers the issue:
- `Agda.Builtin.Unit`(short: `U`) defines the unit `Top` type with a Haskell binding.
- modules `A` and `B` both import `U`
- module `A` is compiled first:  After type checking, we have files `{U,A}.agdai`. The compiler compiles `U` to `U.hs`.  In the process, it discovers that `Top` is not erasable.  This information is however _not_ serialized into the `U.agdai` file (the compiler does not modify interface files).  It compiles `A` using the information that `Top` is not erasable.
- module `B` is compiled second.  It discovers that `U.hs` exists and skips the compilation process for `U.agda`.  Thus, it does not discover that `Top` is not erasable.  It compiles `B` treating `Top` as eraseable.

The solution thus seems to be to not skip the erasability info propagation even if we skip compilation.
I'll try this...

**Comment by andreasabel** [comment]
Unfortunately, my attempt 126cab169a4d9e80b46fba102e5c7252720459ca did not fix the problem, we still have:
```diff
diff --git a/tmp/CompileAsPattern22448-0.golden b/tmp/CompileAsPattern22448-1.actual
index 105a09d..8736cec 100644
--- a/tmp/CompileAsPattern22448-0.golden
+++ b/tmp/CompileAsPattern22448-1.actual
@@ -16,8 +16,7 @@ out >               (Agda.Builtin.List.List._∷_ f (CompileAsPattern._.merge _
 out > CompileAsPattern.mapM! =
 out >   λ _ a b →
 out >     case b of
-out >       Agda.Builtin.List.List.[] →
-out >         Common.IO.return () _ Agda.Builtin.Unit.⊤.tt
+out >       Agda.Builtin.List.List.[] → Common.IO.return () _ _
 out >       Agda.Builtin.List.List._∷_ c d →
 out >         Common.IO._>>=_
 out >           () () _ _ (a c) (λ _ → CompileAsPattern.mapM! _ a d)
```
The problem might be that `CompileAsPattern` is not importing `Unit` directly, but via `Common.Prelude`.  Thus, if `Common.Prelude` is already compiled, `Unit` might not be visited still, even if `Common.Prelude` is visited for the non-erasure propagation.

My next try is to propagate the not-erase information in the `preCompile` hook of the `Backend` infrastructure.  I attempt to just walk over the whole signature, including the imported symbols.  I hope the signature is in place here (frankly, I do not fully understand the compilation logic).

**Comment by andreasabel** [comment]
Walking the whole signature seems to have a noticeable performance impact.  
Maybe the problem could be solved much more easily, by just having a flag for the treeless compiler whether it should keep types with a `HaskellPragma`.

**Comment by andreasabel** [comment]
> Walking the whole signature seems to have a noticeable performance impact.

Not sure about what I said here.  The performance degradation of the compiler-test could as well be caused by the diminished erasure.

This does not seem to be the first test failing on travis, but not locally.  I did it before (#2757), commenting out a test, shame on me!
https://github.com/agda/agda/blob/fa8dc45ca933938a316d031d490e8cb90be3ecdb/test/Compiler/Tests.hs#L66-L77

**Comment by andreasabel** [comment]
I think we have been chasing a phantom here.  The travis failure seems to be that the two compilers MAlonzo (for GHC) and JS (for node) now behave _differently_ wrt. to the debug output produced by test case `CompileAsPattern`, but there is a _single_ golden value they are supposed to match.

I have disabled the `MAlonzo` version of `CompileAsPattern`, while in fact the `JS` version is now deviating from the golden value (after I changed the golden value but not the JS compiler).

With MAlonzo and JS compilers now producing already different Treeless intermediate code, we cannot have Treeless debug output in golden values, unless we decouple the golden values of MAlonzo and JS.

**Comment by rwe** [comment]
@andreasabel
> The travis failure seems to be that the two compilers MAlonzo (for GHC) and JS (for node) now behave differently wrt. to the debug output produced by test case CompileAsPattern,

This does not match up with my experiments. See my most recent comments: the difference was just whether compiled output files already existed. I was reproducing by running Agda manually and did not touch anything to do with JS.

**Comment by rwe** [comment]
Oh, but I think I misunderstood what you were saying. You seem to be saying you fixed the issue with compilation to Haskell but the tests are failing now for a different reason, that the output differs between HS/JS. Sorry, that makes sense.

**Comment by nad** [comment]
If I understand the discussion correctly the potential problem that you discussed [above](https://github.com/agda/agda/issues/3732#issuecomment-489917081) did not trigger this bug. However, it might still be or have been a problem. Was it a problem? And did you fix it? If it is a problem that has not been fixed, then I think we should open a new ticket for it.

**Comment by andreasabel** [comment]
The problem https://github.com/agda/agda/issues/3732#issuecomment-489917081 (which was indeed a problem, thanks for clarifying this, @erydo) was fixed, first in commit e160bf18e94917f82f005788dce6247827cad353 and then, in a more modular way, in commit 9ab7b474c15b6c2301754eb712bceaaaefeaceb3.  There is no further action needed on this.

There is however still the problem that the two compilers GHC and JS share the same golden value (https://github.com/agda/agda/issues/3732#issuecomment-490096983), thus, as they now differ in Treeless debug output which ends up in the golden value, we might want to allow backend-specific golden values for the same test case. 
Then we can bring the Treeless output back which was removed in commit ca3f66da8dd68177da6535e8dc67ef4d14942ce8.
